### ISSUE TITLE
[DocDB] flaky test: MasterTest.TestRegisterDistBroadcastDupPrivateUsePrivateIpNeverCheckUsed

### ISSUE THREAD

**Issue by bmatican** [yugabyte/yugabyte-db#17898]
Jira Link: [DB-6980](https://yugabyte.atlassian.net/browse/DB-6980)
### Description

https://detective-gcp.dev.yugabyte.com/stability/test?analyze_trends=true&branch=master&build_type=all&class=MasterTest&fail_tag=all&name=TestRegisterDistBroadcastDupPrivateUsePrivateIpNeverCheckUsed&platform=linux

seems only flaky in 

alma8-gcc11-fastdebug
--

<br class="Apple-interchange-newline">

### Warning: Please confirm that this issue does not contain any sensitive information

- [X] I confirm this issue does not contain any sensitive information.

[DB-6980]: https://yugabyte.atlassian.net/browse/DB-6980?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**Comment by rthallamko3** [comment]
@dr0pdb , Should we disable the test in fastdebug and unmute the test?

**Comment by dr0pdb** [comment]
@rthallamko3 Sounds good. I'll take a look tomorrow IST time and disable and unmute if the fix isn't simple. Haven't had the chance to investigate.

**Comment by dr0pdb** [comment]
I investigated this a bit. It is not related to this particular test.

The tests fail with the following error messages:
```
F20230831 04:58:56 ../../src/yb/rpc/messenger.cc:225] Scheduled tasks is not empty after messenger shutdown: [{7, 0x000035debed1f600 -> { id: 7 source_location: ../../src/yb/master/async_rpc_tasks.cc:443 }}]
...
```

This is a race condition between scheduling of tasks and messenger shutdown at the end of the tests. It was attempted to be fixed in https://github.com/yugabyte/yugabyte-db/issues/17798 but it is still happening. Seems also related to https://github.com/yugabyte/yugabyte-db/issues/15945.

Before scheduling tasks, we do check if the messenger is shutting down but there is quite a bit of logic in between the check for shutdown and actual insertion into the `scheduled_tasks_` map. The shutdown can happen in between which leads to 

In the failure scenario, the following sequence of operations happen:
1. Thread 1: Before scheduling a task, [we check if the messenger is closing](https://github.com/yugabyte/yugabyte-db/blob/fa6b2b3c2e20d8ff307a11f26fa96ee48356de6e/src/yb/rpc/messenger.cc#L695). This check passes
2. Thread 2: Starts shutdown of messenger
3. Thread 1: Inserts the task into the `scheduled_tasks_` assuming that the messenger is not shutting down since it checked in step 1.

If I move the check for `closing_` [just before the addition of task](https://github.com/yugabyte/yugabyte-db/blob/fa6b2b3c2e20d8ff307a11f26fa96ee48356de6e/src/yb/rpc/messenger.cc#L718) into `scheduled_tasks_`, the test is no longer flaky. 

Basically doing this helps fix the flakiness.
```cpp
  {
    std::lock_guard guard(mutex_scheduled_tasks_);
    if (closing_.load(std::memory_order_acquire)) { // Check right before inserting.
      return STATUS(Aborted, "Cannot schedule task, messenger is closing");
    }
    scheduled_tasks_.emplace(task_id, task);
  }
```

It is still not guaranteed to fix the race condition since the shutdown can happen in between the check and the insert. So we still need to think a bit more on a correct solution.
